### PR TITLE
Adds emojis command

### LINKED ISSUE #80: Emoji list command

### PR DESCRIPTION
**Pull request by moruzerinho6**
image

This closes #80

### CODE REVIEW

File: src/locales/en-US/commands.json [hunk 1]
```diff
@@ -1574,5 +1574,10 @@
     "commandUsage": "<tracking code>",
     "lastUpdated": "Last Updated at",
     "notFound": "Code not found"
+  },
+  "emojis": {
+    "commandDescription": "Take a look at all emojis in the server.",
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:1579]:
```suggestion
    "commandDescription": "Lists all emoji within this server",
```

File: src/locales/en-US/commands.json [hunk 2]
```diff
@@ -1574,5 +1574,10 @@
...
+    "loading": "Loading emojis...",
+    "guildHasNoEmoji": "This guild has no emoji or all emojis are not available for me."
```
Comment (pedrofracassi) [inline src/locales/en-US/commands.json:1581]:
```suggestion
    "guildHasNoEmoji": "This server has no emoji or all emojis are not available for me."
```

File: src/locales/en-US/errors.json [hunk 3]
```diff
@@ -47,5 +47,6 @@
   "validLinkButGhostChannel": "Your link is valid but the bot couldn't find that channel in this guild.",
   "validLinkButGhostMessage": "Your link is valid but I couldn't find that message.",
   "invalidMessageLink": "The message link you sent is invalid.",
-  "messageContainsNothing": "The message you're trying to move has no content."
+  "messageContainsNothing": "The message you're trying to move has no content.",
+  "guildHasNoEmoji": "This guild has no emoji or all emojis are not available for me."
```
Comment (pedrofracassi) [inline src/locales/en-US/errors.json:51]:
```suggestion
  "guildHasNoEmoji": "This server has no emoji or all emojis are not available for me."
```

**Comment by moruzerinho6** [comment]
I need help to remove the loading from the message after sending the embed, I don't know how.

**Comment by davipatury** [comment]
You could use reduce/map to increase the code readability.

```js
const pagesArr = message.guild.emojis.cache
  .filter(emoji => emoji.available)
  .reduce((descriptionArr, emoji, id) => {
    const line = `${emoji.toString()} **${emoji.name}** \`\`<:${emoji.name}:${id}>\`\``
    const [ lastDescription ] = descriptionArr.slice(-1)
    if (lastDescription && ([ ...lastDescription, line]).join('\n').length < 2048) {
      lastDescription.push(line)
    } else {
      descriptionArr.push([ line ])
    }
    return descriptionArr
  }, [])
  .map(desc => new SwitchbladeEmbed(author).setDescription(desc.join('\n')))
    
if (pagesArr.length < 0) {
  throw new CommandError(t('errors:guildHasNoEmoji'))
}

const pages = new PaginatedEmbed(t, author, pagesArr)
try {
  const msg = await channel.send(t('commands:emojilist.loading'))
  pages.run(msg)
} catch (e) {
  throw new CommandError(t('commands:move.couldntSendMessage'))
}
```

**Comment by davipatury** [comment]
You could also improve `PaginatedEmbed.run()` method to support text channels as the first parameter: ```js
async run (target, options = {}) {
  this._footer()
  if (!options.filter) options.filter = () => true
  const emojis = this._determineEmojis(
    [],
    validateProp(options.stop),
    validateProp(options.jump),
    validateProp(options.firstLast)
  )

  let msg
  if (target instanceof TextChannel || target instanceof DMChannel) {
    msg = await target.send(this.pages[options.startPage || 0])
  } else if (target instanceof Message) {
    msg = await target.edit({ embed: this.pages[options.startPage || 0] })
  } else {
    throw new TypeError('Invalid "target" parameter type, must be "TextChannel", "DMChannel" or "Message".')
  }

  return new ReactionHandler(
    msg,
    (reaction, user) => emojis.includes(reaction.emoji.id || reaction.emoji.name) && user !== target.client.user && user === this.author && options.filter(reaction, user),
    options,
    this,
    emojis
  )
}
```

File: src/commands/utility/emojilist.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        if (lastDescription && ([...lastDescription, line]).join('\n').length < 2048) {
+          lastDescription.push(line)
+        } else {
+          descriptionArr.push([line])
+        }
+        return descriptionArr
+      }, [])
+      .map(desc => new SwitchbladeEmbed(author).setDescription(desc.join('\n')))
+
+    if (pages.length < 0) {
```
Comment (pedrofracassi) [inline src/commands/utility/emojilist.js:28]: Wouldn't it be wiser to check if the guild has emoji _before_ generating the embed instead?

Comment (pedrofracassi) [reply src/commands/utility/emojilist.js:28]: Also, will this check ever work? Can the length be less than zero?

**Comment by pedrofracassi** [comment]
I got a bunch of ratelimit log lines while running the command, is this supposed to be happening?

Code_B6yxMIjrxA

**Comment by moruzerinho6** [comment]
> I got a bunch of ratelimit log lines while running the command, is this supposed to be happening?
> 
> Code_B6yxMIjrxA

Are you running in debug mode? @jaxxibae Said that it's expected.

File: src/commands/utility/emojilist.js (hunk 4 above)
Comment (moruzerinho6) [reply src/commands/utility/emojilist.js:28]: Nope, gonna fix when I can.

**Comment by jaxxibae** [comment]
It seems to be debug mode only, and that's just to say when does the bot react next, which in this case, is in 750ms. No need to panic, it's not an error.

**Comment by pedrofracassi** [comment]
Cool. No worries then.

File: src/commands/utility/emojilist.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+      aliases: ['emojis', 'emojislist', 'guildemojis', 'serveremojis'],
+      category: 'utility',
+      requirements: { guildOnly: true, botPermissions: ['ADD_REACTIONS', 'EMBED_LINKS'] }
+    }, client)
+  }
+
+  async run ({ message, t, author, channel }) {
+    const pages = message.guild.emojis.cache
+      .filter(emoji => emoji.available)
+      .reduce((descriptionArr, emoji, id) => {
```
Comment (fernando457829) [inline src/commands/utility/emojilist.js:16]: Stop using `reduce`, it's very confusing. You can do same with:

```js
const pages = [];
const emojis = message.guild.emojis.cache.filter(emoji => emoji.available)

emojis.forEach((emoji) => {
  const line = `${emoji.toString()} **${emoji.name}** \`\`<:${emoji.name}:${id}>\`\``

  const [lastPage] = pages.slice(-1)

  if (lastPage && [...lastPage, line].join('\n').length < 2048) lastPage.push(line)
  else pages.push([line]) 
});
```

File: src/commands/utility/emojilist.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        const line = `${emoji.toString()} **${emoji.name}** \`\`<:${emoji.name}:${id}>\`\``
+        const [lastDescription] = descriptionArr.slice(-1)
+        if (lastDescription && ([...lastDescription, line]).join('\n').length < 2048) {
```
Comment (fernando457829) [inline src/commands/utility/emojilist.js:19]: Unnecessary parentheses
```suggestion
        if (lastDescription && [...lastDescription, line].join('\n').length < 2048) {
```

File: src/commands/utility/emojilist.js (hunk 5 above)
Comment (moruzerinho6) [reply src/commands/utility/emojilist.js:16]: Not applying this change because it was said to not do it while having an internal talk with the team.

**Review by fernando457829 — DISMISSED** [review]

**Comment by fernando457829** [comment]
LGTM

**Review by pedrofracassi — APPROVED** [review]
LGTM!

**Review by pedrofracassi — DISMISSED** [review]

**Review by perronosaurio — APPROVED** [review]

**Review by pedrofracassi — APPROVED** [review]

**Pull request merged**